File: solutions/go/list-ops/1/list_ops.go
package listops

// IntList is an abstraction of a list of integers which we can define methods on
type IntList []int

func (s IntList) Foldl(fn func(int, int) int, initial int) int {
	var result int = initial
	for _, value := range s {
		result = fn(result, value)
	}
	return result
}

func (s IntList) Foldr(fn func(int, int) int, initial int) int {
	// Implementing Foldr
	var result int = initial
	for _, value := range s.Reverse() {
		result = fn(value, result)
	}
	return result
}

func (s IntList) Filter(fn func(int) bool) IntList {
	// Implementing Filter using Foldl
	var result IntList = make([]int, 0)
	for _, value := range s {
		if fn(value) {
			result = append(result, value)
		}
	}
	return result
}

func (s IntList) Map(fn func(int) int) IntList {
	// Implementing Map using Foldl
	var result IntList = make([]int, 0)
	for _, value := range s {
		result = append(result, fn(value))
	}
	return result
}

func (s IntList) Reverse() IntList {
	// Implementing Reverse
	var result IntList = make([]int, 0)
	for _, value := range s {
		result = append(IntList{value}, result...)
	}
	return result
}

func (s IntList) Append(lst IntList) IntList {
	// Implementing Append
	var result IntList = make([]int, 0)
	for _, value := range s {
		result = append(result, value)
	}
	for _, value := range lst {
		result = append(result, value)
	}
	return result
}

func (s IntList) Concat(lists []IntList) IntList {
	// Implementing Concat
	var result IntList = make([]int, 0)
	for _, value := range s {
		result = append(result, value)
	}
	for _, list := range lists {
		for _, value := range list {
			result = append(result, value)
		}
	}
	return result
}

func (s IntList) Length() int {
	var result int = 0
	for range s {
		result++
	}
	return result
}

File: solutions/c/list-ops/1/list_ops.c
#include "list_ops.h"
#include <stdlib.h>
#include <string.h>

list_t *new_list(size_t length, list_element_t elements[]) {
    list_t *list = (list_t *)malloc(sizeof(list_t) + length * sizeof(list_element_t));
    list->length = length;
    if (elements != NULL) {
        memcpy(list->elements, elements, length * sizeof(list_element_t));
    }
    return list;
}

void delete_list(list_t *list) {
    free(list);
}

size_t length_list(list_t *list) {
    return list->length;
}


// append entries to a list and return the new list
list_t *append_list(list_t *list1, list_t *list2) {
    list_t *result = new_list(list1->length + list2->length, NULL);
    memcpy(result->elements, list1->elements, list1->length * sizeof(list_element_t));
    memcpy(result->elements + list1->length, list2->elements, list2->length * sizeof(list_element_t));
    return result;
}

// filter list returning only values that satisfy the filter function
list_t *filter_list(list_t *list, bool (*filter)(list_element_t)) {
    list_t *result = new_list(list->length, NULL);
    size_t result_length = 0;
    for (size_t i = 0; i < list->length; i++) {
        if (filter(list->elements[i])) {
            result->elements[result_length++] = list->elements[i];
        }
    }
    result->length = result_length;
    return result;
}

// return a list of elements whose values equal the list value transformed by
// the mapping function
list_t *map_list(list_t *list, list_element_t (*map)(list_element_t)) {
    list_t *result = new_list(list->length, NULL);
    for (size_t i = 0; i < list->length; i++) {
        result->elements[i] = map(list->elements[i]);
    }
    return result;
}

// folds (reduces) the given list from the left with a function
list_element_t foldl_list(list_t *list, list_element_t initial,
                          list_element_t (*foldl)(list_element_t,
                                                  list_element_t)) {
    list_element_t result = initial;
    for (size_t i = 0; i < list->length; i++) {
        result = foldl(result, list->elements[i]);
    }
    return result;
}

// folds (reduces) the given list from the right with a function
list_element_t foldr_list(list_t *list, list_element_t initial,
                          list_element_t (*foldr)(list_element_t,
                                                  list_element_t)) {
    list_element_t result = initial;
    for (size_t i = list->length; i > 0; i--) {
        result = foldr(list->elements[i - 1], result);
    }
    return result;
}

// reverse the elements of the list
list_t *reverse_list(list_t *list) {
    list_t *result = new_list(list->length, NULL);
    for (size_t i = 0; i < list->length; i++) {
        result->elements[i] = list->elements[list->length - i - 1];
    }
    return result;
}
